feat/195: trust REMOTE_ADDR header only

Sidsector9 opened the pull request:
<!--
### Requirements

Filling out the template is required. Any pull request that does not include enough information to be reviewed in a timely manner may be closed at the maintainers' discretion.  All new code requires documentation and tests to ensure against regressions.
-->

### Description of the Change

This PR proposes a solution to prevent IP spoofing by trusting only the `REMOTE_ADDR` header.
It adds 2 new filters to filter trusted proxies and; headers that the trusted proxies use to forward client IP information.

### Alternate Designs

TBD.

### Possible Drawbacks

This will be a breaking change.

### Verification Process

<!--
What process did you follow to verify that your change has the desired effects?

- How did you verify that all new functionality works as expected?
- How did you verify that all changed functionality works as expected?
- How did you verify that the change has not introduced any regressions?

Describe the actions you performed (e.g., commands you ran, text you typed, buttons you clicked) and describe the results you observed.
-->

### Checklist:

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] I have read the [**CONTRIBUTING**](https://github.com/10up/restricted-site-access/blob/develop/CONTRIBUTING.md) document.
- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my change.
- [ ] All new and existing tests passed.

<!-- _NOTE: these things are not required to open a PR and can be done afterwards / while the PR is open._ -->

### Applicable Issues

Closes #195 

### Changelog

> Security - New filters - `rsa_trusted_proxies` and `rsa_trusted_headers` have been added to help prevent IP spoofing attacks.

### Credits

@Sidsector9, @dkotter, @peterwilsoncc, @marcS0H, @DanielRuf.

Sidsector9 committed:
try: trust REMOTE_ADDR header only

jeffpaul commented:
@Sidsector9 @dkotter what else is needed to get this PR open for review (and then merged and ready for release)?

dkotter commented on restricted_site_access.php:
```diff
@@ -1529,36 +1529,85 @@ public static function ip_in_range( $ip, $range ) {
+		 *
+		 * @since 7.3.1
+		 */
+		$trusted_proxies = apply_filters( 'rsa_trusted_proxies', 'REMOTE_ADDR' );
```
If I'm following things correctly, the idea is for this value to be an array of IP addresses that correspond to the proxy server a site is running behind. In that case, I don't think we should set `REMOTE_ADDR` as the default here. I think the default should just be `false` or an empty string. This then would change some of the logic below as well.

dkotter commented on restricted_site_access.php:
```diff
@@ -1529,36 +1529,85 @@ public static function ip_in_range( $ip, $range ) {
+		 */
+		$proxy_trusted_headers = apply_filters( 'rsa_proxy_trusted_headers', array( 'HTTP_X_FORWARDED_FOR' ) );
+	
+		if ( is_string( $trusted_proxies ) && 'REMOTE_ADDR' === $trusted_proxies ) {
```
Going off my comment above, I'd change this check to look for an empty array or empty string, since we technically don't ever utilize the value `REMOTE_ADDR`

dkotter commented on restricted_site_access.php:
```diff
@@ -1529,36 +1529,85 @@ public static function ip_in_range( $ip, $range ) {
+		}
+	
+		/**
+		 * Accepts the string 'REMOTE_ADDR' or array of trusted proxies.
+		 * It is advisable to pass 'REMOTE_ADDR' if your proxy server doesn't have
+		 * a static IP.
```
If my assumption is correct that this should be utilized to add trusted IP addresses, I think we need to change the comment here to make that clear. Right now, this is confusing to me

Sidsector9 replied:
> If my assumption is correct that this should be utilized to add trusted IP addresses

Correct. If the array is not set, then the default string `REMOTE_ADDR` is used. The situation where you won't set this array is when the proxy has dynamic IPs. in that case we trust any value that is set in `REMOTE_ADDR`

dkotter commented on restricted_site_access.php:
```diff
@@ -1529,36 +1529,85 @@ public static function ip_in_range( $ip, $range ) {
+		 *
+		 * @since 7.3.1
+		 */
+		$proxy_trusted_headers = apply_filters( 'rsa_proxy_trusted_headers', array( 'HTTP_X_FORWARDED_FOR' ) );
```
So if neither of these new filters are utilized, the end result is we only look at the `HTTP_X_FORWARDED_FOR` header, is that correct? For security purposes, I feel we should only look at the `REMOTE_ADDR` header and then these filters can be utilized to add additional header support

Sidsector9 replied:
Yes that is how I've set the default behaviour as it is the most common pattern followed. I will change this.

dkotter commented:
> what else is needed to get this PR open for review (and then merged and ready for release)?

I've left some comments on things I think should be changed, though I haven't tested this out yet.

I do think it's still worth discussing on if this is something we want to fix and if this is the best way to fix it. My concern still stands that this is a backwards compatibility break and I'm not super comfortable with that. I like this new approach of handling things as it is more secure but in essence this will potentially break (meaning not work as expected) any site that is running behind some sort of proxy (like a CDN). I almost like the idea of having this be opt in behavior, either through a filter or a new setting. So by default we keep things as-is and if individual sites want better security, they can turn this new approach on.

jeffpaul commented:
New opt-in filter/setting seems like a good approach as well, in no way looking to introduce a breaking change without a user taking an educated action to trigger that change.

jeffpaul commented:
@Sidsector9 checking back in to see if you're able to get this ready for review/merge?

jeffpaul committed:
Merge branch 'develop' into fix/195

Sidsector9 commented:
@jeffpaul I'll go through Darin's comments and then estimate how much time it'll take.

Sidsector9 committed:
refactor logic

Sidsector9 commented:
@dkotter I've responded to your comments and refactored the logic a bit and updated the comments with more details. The code reflects the below flowchart.

If this gets approved, then I'll proceed with adding this as an opt-in feature.

<img width="1016" alt="Screenshot 2022-07-11 at 1 37 11 PM" src="https://user-images.githubusercontent.com/17757960/178252327-0e928333-4e34-4fec-af91-aa7bcbe9ad4d.png">

Sidsector9 committed:
Merge branch 'develop' into fix/195

dkotter committed:
Iterate over each trusted proxy IP address and pass into our validation function, instead of doing a strict match. Set the default for our trusted headers to match the current functionality to avoid backwards compat issues. Documentation changes

dkotter commented:
Thanks for all the work here @Sidsector9 (as well as continuing to remind me this needed my attention). I've made a few minor changes, mostly around maintaining backwards compatibility with the existing list of headers we support by default.

For information sake, the approach here (or at least the approach I was looking at, don't want to speak for Sid) was based on how Symfony handles proxies (see [docs](https://symfony.com/doc/current/deployment/proxies.html), [code](https://github.com/symfony/symfony/blob/3.2/src/Symfony/Component/HttpFoundation/Request.php#L803)).

Here's an overview of how this should be working now:

1. We get the IP address from the `REMOTE_ADDR` header. If that doesn't exist, we return early with an empty string
2. A new filter is run that can be used to set the trusted IP addresses of a proxy. By default this returns an empty array, so user action is needed here to set this up.
3. If any value is added via the filter, we then validate the `REMOTE_ADDR` value matches the value of the trusted proxy IP. If it doesn't match, we return early with an empty string
4. If it does match or no trusted proxy IP addresses were added, we then get a list of trusted HTTP headers we want to look at. By default this is the exact same as it is now but a filter is added around this that can be used to add or remove trusted headers. The recommendation would be for any sites not behind a proxy, use this new filter to remove all headers (return an empty array)
5. We then add the `REMOTE_ADDR` header to this list so that header will always be checked
6. All of those headers are then iterated through and as soon as we find a match, we return that

There are two things I'd love some feedback on here (cc/ @jeffpaul @peterwilsoncc):

1. This new functionality is all behind filters, so code-level changes are needed to take advantage of this. We could add new settings to our admin UI to handle this but I'd recommend either not doing that at all or doing that in a later iteration (concern with cluttering that UI too much and also want to get this PR out sooner)
2. The changes we've made were done with backwards compatibility in mind. This means out of the box, things aren't any more secure than they are now, the new filters will need to be utilized to take advantage of that. I'm on the fence on this one but we could do a major version bump here and change this so we're more secure by default and you'd have to use these new filters to get back to where we are now. That impacts all existing users but does force all users to be more secure

dkotter committed:
Add back header

dkotter commented:
~~Note (mostly for myself so I don't forget) I see two unit tests are failing here. At first glance I think this isn't a regression in the code changes here but actually highlights a potential bug that is now fixed.~~

~~If you look at the current code that determines the IP address, we set the `$ip` variable to an empty string [here](https://github.com/10up/restricted-site-access/blob/develop/restricted_site_access.php#L1532). We then iterate through all our headers and set each of those to the variable `$ip` [here](https://github.com/10up/restricted-site-access/blob/develop/restricted_site_access.php#L1552). But we only return that value if it passes the `filter_var` check [here](https://github.com/10up/restricted-site-access/blob/develop/restricted_site_access.php#L1558). And then at the very end of this function we [return](https://github.com/10up/restricted-site-access/blob/develop/restricted_site_access.php#L1564) the `$ip` variable.~~

~~So if none of the IP addresses validate, instead of returning an empty string we return the last value that was stored in `$ip`, which will be the last header that we process (should always be `REMOTE_ADDR`). In the case of our tests, the IP address is 127.0.0.1, which doesn't pass the `filter_var` check. In the current code, it still ends up returning that value but in the changes we've made here, it will return an empty string. I think that's the right approach but maybe we should always return the IP address stored in `REMOTE_ADDR`? In real world usage, that should always validate and shouldn't be a problem, but just need to decide for the sake of those tests.~~

EDIT: I've left the above for posterity but I've gone ahead and made sure we try and always have a default IP that we return (this will be the `REMOTE_ADDR` header). I can still see this being a bug but I can also see it being intended behavior and it is how it works now. This fixes the tests

peterwilsoncc commented on restricted_site_access.php:
```diff
@@ -1529,26 +1529,98 @@ public static function ip_in_range( $ip, $range ) {
+			return '';
+		}
+
+		/*
```
```suggestion
		/**
```

Docblock

peterwilsoncc commented on restricted_site_access.php:
```diff
@@ -1529,26 +1529,98 @@ public static function ip_in_range( $ip, $range ) {
-		foreach ( $headers as $key ) {
 
-			if ( ! isset( $_SERVER[ $key ] ) ) {
+		/*
```
```suggestion
		/**
```

Docblock

dkotter committed:
Always return the REMOTE_ADDR header value if nothing else matches

dkotter committed:
Merge branch 'develop' into fix/195

dkotter committed:
Make sure we actually return our variable

dkotter committed:
Add tests

Set our REMOTE_ADDR header in our tests

Properly reset things between tests

Change filter flags so local IPs pass verification

Move filter location

peterwilsoncc commented on restricted_site_access.php:
```diff
@@ -1529,26 +1529,99 @@ public static function ip_in_range( $ip, $range ) {
+		 */
+		$trusted_proxies = apply_filters( 'rsa_trusted_proxies', array() );
+
+		if ( ! empty( $trusted_proxies ) ) {
```
The `empty()` prevents developers disallowing any proxies. I'd expect the following to prevent proxy access:

```php
add_filter( 'rsa_trusted_proxies', '__return_empty_array' );
```

It might be worht checking with `has_filter()` but I could be convinced otherwise.

dkotter replied:
Just want to make sure I'm following correctly. You're suggesting we replace the `empty` check with a `has_filter` check, allowing someone to use this filter to return an empty array, which would then pass our check but because there's no proxies to iterate over at that point, we would just return an empty string which should in essence block access to the site?

Guess I'm just wondering what the use case here would be for this, seems like this could be achieved already using the settings in the admin. I'm not opposed to making this change just having a hard time coming up with a scenario where this would be helpful.

peterwilsoncc replied:
It's possible I am misunderstanding the purpose of the code. Let me do some testing during the day and get back to you. If needs be I will provide a test to clarify my thinking.

peterwilsoncc replied:
I had myself totally confused when I read this code yesterday. I see what it's doing and it all makes sense to me.

peterwilsoncc commented on tests/php/test-ip-addresses.php:
```diff
@@ -71,4 +71,93 @@ public function test_get_client_ip_address() {
+
+		$this->assertSame( '127.0.0.1', $rsa::get_client_ip_address() );
+
+		// Reset the filter.
+		add_filter( 'rsa_trusted_proxies', '__return_empty_array' );
```
```suggestion
```

Redundant, `WP_UnitTestCase` resets the filters between each test. ([source ref](https://github.com/WordPress/wordpress-develop/blob/13e076e6a48deb388bcba3a23beea2a2465a82d5/tests/phpunit/includes/abstract-testcase.php#L194))

peterwilsoncc reviewed (commented):
A few notes inline.

It would be nice if the tests used data providers rather than multiple tests in a single block, but not a blocker.

dkotter committed:
CR feedback

dkotter committed:
Modify tests a bit to use data providers

Fix our tests to ensure we set all headers first before asserting the value

More test clean up

More test clean up

dkotter committed:
Ensure we have an IP before checking it

Sidsector9 commented:
@jeffpaul this is approved. As discussed yesterday in 1:1, this can go with 7.3.2.

dkotter committed:
Update readmes with information on how to utilize these new filters

dkotter commented:
Note that I've updated the readmes with information on how to utilize these new filters

jeffpaul commented:
@Sidsector9 can you please update the changelog entry in the PR description (I added the credit line folks), thanks!
